src/rgw/rgw_common.cc

```diff
@@ -1467,3 +1478,78 @@ int rgw_parse_op_type_list(const string& str, uint32_t *perm)
   return parse_list_of_flags(op_type_mapping, str, perm);
 }
 
+static int match_internal(const char* pattern, const char* input)
+{
+  while(true) {
+    if (*pattern == '\0' && *input == '\0')
+      return 1;
+    if (*pattern == '\0' || *input == '\0')
+      return 0;
+    if (*pattern == '*' && *(pattern + 1) == '\0' && *input != '\0')
+      return 1;
+    if (*pattern == '*' && *(pattern + 1) == '\0' && *input == '\0')
+      return 0;
+    if (*pattern == *input || *pattern == '?') {
+      pattern++;
+      input++;
+      continue;
+    }
+    if (*pattern == '*') {
+      if (*(pattern + 1) == *input)
+        pattern++;
+      else
+        input++;
+      continue;
+    }
+    return 0;
+  }
+  return 0;
+}
+
+static void tokenize(string& text, vector<string>& result)
+{
+  char_separator<char> sep(":");
+  tokenizer<char_separator<char> > tokens(text, sep);
+  for ( auto it : tokens)
+  {
+    result.push_back(it);
+  }
+}
+
+int match(string& pattern, string& input, int flag)
+{
+  if (flag & POLICY_ACTION || flag & POLICY_ARN) {
+    std::transform(pattern.begin(),
+                    pattern.end(),
+                    pattern.begin(),
+                    ::tolower);
+    std::transform(input.begin(),
+                   input.end(),
+                   input.begin(),
+                   ::tolower);
+  }
+  vector<string> result1, result2;
+  tokenize(pattern, result1);
+  tokenize(input, result2);
+
+  if (result1.size() != result2.size()) {
+      return 0;
+  }
+
+  if (flag & POLICY_ACTION) {
+      vector<string> services = {"iam", "ec2", "qs", "sns", "s3"};
+      if (std::find(std::begin(services), std::end(services), result1[0])
+          == std::end(services)) {
+          return 0;
+      }
```

Comment:
Yes, the document says that in case of ACTIONS, the ARNs should begin with a valid service. Though I am not sure if this list is complete. I'll have to search more.